src/sentry/sentry_metrics/indexer/base.py

```diff
@@ -1,8 +1,162 @@
+
+        return key_pairs
+
+    def as_strings(self) -> Sequence[str]:
+        """
+        Returns all the keys, each key represented as string -> "1:a"
+        """
+        keys: MutableSequence[str] = []
+        for org_id in self.mapping:
+            keys.extend([f"{org_id}:{string}" for string in self.mapping[org_id]])
+
+        return keys
+
+
+class KeyResults:
+    def __init__(self) -> None:
+        self.results: MutableMapping[int, MutableMapping[str, int]] = defaultdict(dict)
+        self.meta: MutableMapping[FetchType, MutableMapping[int, str]] = defaultdict(dict)
+
+    def add_key_result(self, key_result: KeyResult, fetch_type: Optional[FetchType] = None) -> None:
+        self.results[key_result.org_id].update({key_result.string: key_result.id})
+        if fetch_type:
+            self.meta[fetch_type].update({key_result.id: key_result.string})
+
+    def add_key_results(
+        self, key_results: Sequence[KeyResult], fetch_type: Optional[FetchType] = None
+    ) -> None:
+        for key_result in key_results:
+            self.results[key_result.org_id].update({key_result.string: key_result.id})
+            if fetch_type:
+                self.meta[fetch_type].update({key_result.id: key_result.string})
+
+    def get_mapped_results(self) -> Mapping[int, Mapping[str, int]]:
+        """
+        Only return results that have org_ids with string/int mappings.
+        """
+        mapped_results = {k: v for k, v in self.results.items() if len(v) > 0}
+        return mapped_results
+
+    def get_unmapped_keys(self, keys: KeyCollection) -> KeyCollection:
+        """
+        Takes a KeyCollection and compares it to the results. Returns
+        a new KeyCollection for any keys that don't have corresponding
+        ids in results.
+        """
+        unmapped_org_strings: MutableMapping[int, Set[str]] = defaultdict(set)
+        for org_id, strings in keys.mapping.items():
+            for string in strings:
+                if not self.results[org_id].get(string):
+                    unmapped_org_strings[org_id].add(string)
+
+        return KeyCollection(unmapped_org_strings)
+
+    def get_mapped_key_strings_to_ints(self) -> MutableMapping[str, int]:
+        """
+        Return the results, but formatted as the following:
+            {
+                "1:a": 10,
+                "1:b": 11,
+                "1:c", 12,
+                "2:e": 13
+            }
+        This is for when we use indexer_cache.set_many()
+        """
+        cache_key_results: MutableMapping[str, int] = {}
+        for org_id, result_dict in self.results.items():
+            for string, id in result_dict.items():
+                key = f"{org_id}:{string}"
+                cache_key_results[key] = id
+
+        return cache_key_results
+
+    def get_fetch_metadata(self) -> Mapping[FetchType, Mapping[int, str]]:
+        return self.meta
+
+    def merge(self, other: "KeyResults") -> "KeyResults":
```

Comment:
I'd be open to making this an immutable method (e.g. just create a new KeyResults objects with all elements from the instance plus other).

Reply:
Not sure why I did it this way. I think it could arguably perform better if you had thousands or hundreds of thousands of `KeyResults` in memory at a given time, but object creation is probably orders of magnitude faster than Memcache/Postgres fetches.

I think I'll just change this to be immutable